### PR TITLE
DBCP support

### PR DESCRIPTION
**Pull request by gytis**
@zhfeng would you mind reviewing this PR? I've added an extension to add DBCP support based your implementation in Narayana Tomcat. It is configurable and by default still uses Transactional Driver (with GenericXADataSourceWrapper), but adding `narayana.dbcp.enable=true` will switch to using DBCP2.

@geoand I know you're not an expect of Narayana, but could you also skim through this quickly just in case you have any suggestions on autoconfiguration or code structure?

### CODE REVIEW

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java [hunk 1]
```diff
@@ -25,32 +25,37 @@
 import org.springframework.boot.jdbc.XADataSourceWrapper;
 
 /**
- * {@link XADataSourceWrapper} implementation that uses {@link NarayanaDataSource} to wrap an {@link XADataSource}.
+ * An abstract {@link XADataSourceWrapper} implementation which creates a {@link XAResourceRecoveryHelper} capable to
+ * recovery transactions of the provided {@link XADataSource} and then delegates the actual data source wrapper to its
```
Comment (metacosm) [inline narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: This doesn't parse very well…

Comment (geoand) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: I would say it should be `capable of recovering transactions of the provided`

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java [hunk 2]
```diff
@@ -25,32 +25,37 @@
...
-     * @param properties       Narayana properties.
-     * @param xaRecoveryModule recovery module to register data source with.
-     */
-    public NarayanaXADataSourceWrapper(NarayanaProperties properties, XARecoveryModule xaRecoveryModule) {
+    protected AbstractXADataSourceWrapper(NarayanaProperties properties, XARecoveryModule xaRecoveryModule) {
         this.properties = properties;
         this.xaRecoveryModule = xaRecoveryModule;
     }
 
+    protected abstract DataSource internalWrapDataSource(XADataSource dataSource) throws Exception;
```
Comment (geoand) [inline narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: I would usually place the `internal` "qualifier" at the end, but I guess it's just a matter of taste

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java (hunk 1 above)
Comment (gytis) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: Duh :)

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java (hunk 2 above)
Comment (gytis) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: I'll change it

Comment (geoand) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java]: Cool

**Comment by gytis** [comment]
I've updated the javadoc and renamed internalWrapDataSource method

File: pom.xml [hunk 3]
```diff
@@ -53,6 +53,7 @@
     <assertj.version>3.9.0</assertj.version>
     <camel.version>2.20.2</camel.version>
     <checkstyle.version>8.5</checkstyle.version>
+    <commons-dbcp2.version>2.2.0</commons-dbcp2.version>
```
Comment (zhfeng) [inline pom.xml]: It could upgrade to use 2.4.0 which includes the fixes of  DBCP-484 and DBCP-491 related to the Narayana

Comment (gytis) [reply pom.xml]: OK I'll have to test it first. 2.2.0 is the version used in Spring Boot.

**Pull request merged**